packages/core/src/application.ts

```diff
@@ -61,9 +63,83 @@ export class Application extends Context {
    */
   controller(controllerCtor: ControllerClass, name?: string): Binding {
     name = name || controllerCtor.name;
-    return this.bind(`controllers.${name}`)
+    return this.bind(`${CoreBindings.CONTROLLERS_PREFIX}.${name}`)
       .toClass(controllerCtor)
-      .tag('controller');
+      .tag(CoreBindings.CONTROLLERS_TAG);
+  }
+
+  /**
+   * Register a booter class / array of classes with this application.
+   *
+   * @param {Function | Function[]} booterCls The booter class (constructor function).
+   * @param {string=} name Optional booter name, defaults to the class name.
+   * Ignored is cls is an Array and the name defaults to the class name.
+   * @returns {Binding | Binding[]} The newly created binding(s), you can use the
+   * reference to further modify the binding, e.g. lock the value to prevent
+   * further modifications.
+   *
+   * ```ts
+   * class MyBooter implements Booter {}
+   * app.booter(MyBooter);
+   * ```
+   */
+  booter(booterCls: Constructor<Booter>, name?: string): Binding;
+  booter(booterCls: Constructor<Booter>[]): Binding[];
+  booter(
+    booterCls: Constructor<Booter> | Constructor<Booter>[],
+    name?: string,
+    // tslint:disable-next-line:no-any
+  ): any {
+    if (Array.isArray(booterCls)) {
+      return booterCls.map(cls => this._bindBooter(cls));
+    } else {
+      return this._bindBooter(booterCls, name);
+    }
+  }
+
+  /**
+   *
+   * @param booterCls A Booter Class
+   * @param {string} name Name the Booter Class should be bound to
+   * @returns {Binding} The newly created Binding
+   */
+  private _bindBooter<T extends Booter>(
+    booterCls: Constructor<T>,
+    name?: string,
+  ): Binding {
+    name = name || booterCls.name;
+    return this.bind(`${CoreBindings.BOOTER_PREFIX}.${name}`)
+      .toClass(booterCls)
+      .inScope(BindingScope.CONTEXT)
+      .tag(CoreBindings.BOOTER_TAG);
+  }
+
+  /**
+   * Function is responsible for calling all registered Booter classes that
+   * are bound to the Application instance. Each phase of an instance must
+   * complete before the next phase is started.
+   *
+   * @param {BootExecutionOptions} execOptions Options to control the boot
+   * process for the Application
+   */
+  async boot(execOptions?: BootExecutionOptions): Promise<void> {
+    // Get a instance of the BootStrapper
+    const bootstrapper = await this.get(CoreBindings.BOOTSTRAPPER, {
```

Comment:
> > Please cast the result of `this.get` to the correct type (`Bootstrapper`?). Right now, `bootstrapper` has type `any`, which disables any compiler checks in the rest of this method.

> Casting `this.get` will require import from `@loopback/boot` which will create a circular dependency between `core` and `boot`.

In that case we need to rework our design. The circular dependency between `core` and `boot` is still there even in your current implementation, it's just hidden from the compiler so that it cannot complain about this problem.

See, this is one of the reasons why I was advocating against having the built-in `app.boot` implemented via `@loopback/boot`. I don't want to start this discussion again though (unless you want to).

At minimum, `core` should define an interface describing `BOOTSTRAPPER` API - I think all we need is the `boot` method which accepts bootOptions and execOptions. Then the real bootstrapper implementation in `@loopback/boot` should implement this interface.

I am proposing to call the interface as `Bootstrapper` and the implementation class as `ConventionalBootstrapper`, but feel free to pick better names!

Reply:
I don't agree. As I see it `app.boot()` provides nothing but a sugar method to retrieve the `Bootstraper` via DI and call's it's `boot` method. There isn't much else in terms of implementation here. `bootstrapper.boot()` does all the heavy lifting.

If all we need is an interface in `core` for `Bootstrapper` that is implemented by `boot` for types, I can add that but I don't think it'll be good UX for a user to see `bootstrapper.boot()` retrieval code in CLI generated code ... it's another line to understand for the user looking to get started quickly.

Reply:
My view is that `app.boot()` provides a public API contract that any consumer of an application can rely on. That's the most important part to me.

Another important point for me is to keep boot and core as decoupled as possible. I am using this use case to drive/verify that design: LoopBack users should be able to replace our built-in booter with their custom implementation.

Now if we decide that the default `app.boot` is using a Bootstrapper to run, then we should have a public interface describing what is expected from a (custom) Bootstrapper class to provide so that `app.boot` can use it. That's where the proposed interface helps.

> I don't think it'll be good UX for a user to see bootstrapper.boot() retrieval code in CLI generated code ... it's another line to understand for the user looking to get started quickly.

I have a different opinion here - to me, that generated line of code makes it easier for users to understand how `app.boot()` works under the hood and where/how to start customizing it.

However, I am ok to agree that we disagree on this point and put this discussion on hold for now, so that this pull request can be finally landed.

So, to reiterate, if you can add an interface describing the contract of Bootstrapper class that's required by the default `app.boot()` implementation and fix `app.boot()` to cast the result of `ctx.get()` to this interface, then I am fine here.